activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,27 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def unchanged_and_covered_by_unique_index?(klass, record, attribute)
+        return false if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        return false unless attributes_scope?(record, scope)
+
+        attributes = scope + [attribute.to_s]
+        return false if attributes.any? { |attribute| record.attribute_changed?(attribute) }
+
+        klass.connection.schema_cache.indexes(klass.table_name).any? do |index|
+          index.unique &&
+            index.where.nil? &&
+            index.columns.size <= attributes.size &&
```

Comment:
Hum, actually that `index.columns.size <= attributes.size` condition might be wrong, no?

If we validate uniqueness on `(a)`, and there's a unique index on `(a, b)`, we can't assume it's unique for us. I think we need `index.columns.size == attributes.size`.

Another thing to look at is how to deal with `nil / NULL`. In SQL `NULL != NULL`, so a unique index doesn't prevent multiple `NULL` entries.

Reply:
> If we validate uniqueness on (a), and there's a unique index on (a, b), we can't assume it's unique for us. I think we need index.columns.size == attributes.size.

In your example - yes, we can't assume.
 
But this checks when we validate uniqueness on `(a, b)` and have a unique index on `(a, b)` or `(a)` or `(b)` (basically with exact columns or any combination of the columns of smaller size), then we can use it.

> Another thing to look at is how to deal with nil / NULL. In SQL NULL != NULL, so a unique index doesn't prevent multiple NULL entries.

That's a good point! I think we can just skip this optimization when any of the stored values is `nil` and do the actual db query for checking.

Reply:
> with exact columns or any combination of the columns of smaller size

Ah right 🤦 , I was reading it in the opposite direction.